File: Prism/swift version/Sources/Prism/Tools.swift
import Foundation

// MARK: - Tool Registry

/// All retrieval tools available to the main model. Each tool reads from the
/// app's on‑disk JSON archives and returns a JSON result string.
/// Quality guard checks run automatically in the pre‑pipeline — they are not
/// exposed as tools.
enum ToolRegistry {

    /// The `tools` array sent to DeepSeek's API.
    static let definitions: [ToolDef] = [
        .trackPerson,
        .emotionTimeline,
        .searchChapters,
        .fetchChapterMessages,
        .searchMemory,
        .manageNarrativeTimeline,
    ]

    // MARK: - Memory Search

    @MainActor private static func searchMemory(query: String, store: ChatAgent, limit: Int) -> String {
        let entries = store.searchMemory(query: query, limit: limit)
        let results: [[String: Any]] = entries.map { entry in
            [
                "content": entry.content,
                "keywords": entry.keywords,
                "sourceChapter": entry.sourceChapterTitle,
                "timeSpanStart": entry.timeSpanStart?.ISO8601Format() ?? NSNull(),
                "timeSpanEnd": entry.timeSpanEnd?.ISO8601Format() ?? NSNull(),
                "recallCount": entry.recallCount,
            ]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: results),
              let str = String(data: data, encoding: .utf8) else {
            return "[]"
        }
        return str
    }

    /// Execute a retrieval tool call locally and return the result.
    /// Search tools use Flash reranking for semantic understanding when settings are provided.
    @MainActor
    static func execute(
        name: String,
        arguments: String,
        store: ChatAgent,
        settings: AppSettings? = nil
    ) async -> String {
        let args: [String: String]
        if let data = arguments.data(using: .utf8),
           let decoded = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            args = decoded.reduce(into: [:]) { result, pair in
                if let value = pair.value as? String { result[pair.key] = value }
                else if let value = pair.value as? NSNumber { result[pair.key] = value.stringValue }
            }
        } else {
            print("[ToolRegistry] ⚠ invalid UTF-8 or malformed JSON arguments: \(arguments.prefix(120))")
            args = [:]
        }

        let result: String
        switch name {
        case "track_person":
            result = trackPerson(name: args["name"] ?? "", store: store)
        case "emotion_timeline":
            let n = Int(args["count"] ?? "5") ?? 5
            result = emotionTimeline(count: n, store: store)
        case "search_chapters":
            if let settings {
                let n = Int(args["count"] ?? "5") ?? 5
                let chapters = await store.searchChaptersSemantic(query: args["query"] ?? "", settings: settings, limit: n)
                result = await buildChapterResults(chapters: chapters.map(\.chapter), store: store)
            } else {
                result = searchChapters(query: args["query"] ?? "", store: store)
            }
        case "fetch_chapter_messages":
            result = fetchChapterMessages(args: args, store: store)
        case "search_memory":
            let n = Int(args["count"] ?? "10") ?? 10
            if let settings {
                let entries = await store.searchMemorySemantic(query: args["query"] ?? "", settings: settings, limit: n)
                result = encodeMemoryResults(entries)
            } else {
                result = searchMemory(query: args["query"] ?? "", store: store, limit: n)
            }
        case "manage_narrative_timeline":
            result = manageNarrativeTimeline(args: args, store: store)
        default:
            result = #"{"error":"unknown tool: \#(name)"}"#
        }

        // Log tool execution for observability
        let preview = result.count > 120 ? String(result.prefix(120)) + "…" : result
        print("[Tool] \(name) | args: \(arguments.prefix(80)) | → \(preview)")

        return result
    }

    @MainActor private static func manageNarrativeTimeline(args: [String: String], store: ChatAgent) -> String {
        let action = args["action"] ?? "list"
        if action == "list" {
            let events = store.narrativeEvents(for: store.selectedConversationID)
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            guard let data = try? encoder.encode(events),
                  let string = String(data: data, encoding: .utf8) else { return "[]" }
            return string
        }

        let eventID = args["eventId"].flatMap(UUID.init(uuidString:))
        guard let event = store.upsertNarrativeEvent(
            eventID: eventID,
            title: args["title"] ?? "",
            summary: args["summary"] ?? "",
            startLabel: args["startLabel"] ?? "",
            endLabel: args["endLabel"] ?? "",
            timeKind: args["timeKind"] ?? "period",
            sortIndex: Int(args["sortIndex"] ?? "10") ?? 10
        ) else {
            return #"{"error":"title and a clear time label are required; ask the user to clarify before writing"}"#
        }
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(event),
              let string = String(data: data, encoding: .utf8) else { return #"{"saved":true}"# }
        return #"{"saved":true,"event":\#(string)}"#
    }

    /// Encode memory entries as JSON for tool response.
    @MainActor private static func encodeMemoryResults(_ entries: [MemoryEntry]) -> String {
        let results: [[String: Any]] = entries.map { entry in
            [
                "content": entry.content,
                "keywords": entry.keywords,
                "sourceChapter": entry.sourceChapterTitle,
                "recallCount": entry.recallCount,
            ]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: results),
              let str = String(data: data, encoding: .utf8) else {
            return "[]"
        }
        return str
    }

    // MARK: - Individual Tool Executors

    /// Look up a person across all conversations.
    @MainActor private static func trackPerson(name: String, store: ChatAgent) -> String {
        let archive = store.personArchive
        guard !name.isEmpty else {
            return #"{"found":false,"reason":"empty name"}"#
        }
        guard let person = archive.first(where: { $0.name.localizedCaseInsensitiveContains(name) }) else {
            return #"{"found":false,"name":"\#(name)"}"#
        }

        let encoder = JSONEncoder()
        guard let json = try? encoder.encode(person),
              let str = String(data: json, encoding: .utf8) else {
            return #"{"found":false}"#
        }
        return str
    }

    /// Return recent emotional trajectory — raw data only, model judges the trend.
    @MainActor private static func emotionTimeline(count: Int, store: ChatAgent) -> String {
        let timeline = store.emotionTimeline.suffix(count)
        var result: [[String: Any]] = []
        for entry in timeline {
            result.append([
                "emotion": entry.emotion,
                "intensity": entry.intensity,
                "date": ISO8601DateFormatter().string(from: entry.createdAt),
                "segment": entry.segment,
            ])
        }

        guard let data = try? JSONSerialization.data(withJSONObject: ["entries": result]),
              let str = String(data: data, encoding: .utf8) else {
            return #"{"entries":[]}"#
        }
        return str
    }

    /// Search past chapter summaries — returns full message content for matches
    /// so the agent can read the original conversation text.
    @MainActor private static func searchChapters(query: String, store: ChatAgent) -> String {
        let allChapters = store.allChapters
        let topN = 10

        guard !query.isEmpty else {
            let recent = allChapters.suffix(topN)
            return buildChapterResults(chapters: Array(recent), store: store)
        }

        let q = query.lowercased()
        let terms = expandQuery(q.split(separator: " ").map(String.init).filter { $0.count >= 1 })

        var scored: [(chapter: StoryChapter, score: Int)] = []
        for ch in allChapters {
            let t = ch.title.lowercased()
            let s = ch.summary.lowercased()
            let kw = ch.keywords.map { $0.lowercased() }
            var score = 0

            for term in terms {
                if t.contains(term) { score += 3 }
                else if kw.contains(where: { $0.contains(term) }) { score += 2 }
                if s.contains(term) { score += 1 }
            }
            if t.contains(q) { score += 2 }
            if kw.contains(where: { $0.contains(q) }) { score += 1 }

            if score > 0 { scored.append((ch, score)) }
        }

        scored.sort { $0.score > $1.score }
        let top = Array(scored.prefix(topN).map(\.chapter))

        return buildChapterResults(chapters: top, store: store)
    }

    /// Fetch all messages for a specific chapter by index (1-based).
    @MainActor private static func fetchChapterMessages(args: [String: String], store: ChatAgent) -> String {
        guard let indexStr = args["index"],
              let index = Int(indexStr),
              index >= 1 else {
            return #"{"error":"invalid index"}"#
        }

        let allChapters = store.allChapters
        guard index <= allChapters.count else {
            return #"{"error":"chapter not found","max":\#(allChapters.count)}"#
        }

        let chapter = allChapters[index - 1]
        let msgs = store.messages(for: chapter)
            .filter { $0.role != .system }
            .prefix(12)

        let messageList: [[String: String]] = msgs.map { msg in
            ["role": msg.role.rawValue, "content": msg.content]
        }

        let result: [String: Any] = [
            "chapter": index,
            "title": chapter.title,
            "summary": chapter.summary,
            "keywords": chapter.keywords,
            "messages": messageList,
        ]

        guard let data = try? JSONSerialization.data(withJSONObject: result),
              let str = String(data: data, encoding: .utf8) else {
            return #"{"error":"serialization failed"}"#
        }
        return str
    }

    @MainActor private static func buildChapterResults(chapters: [StoryChapter], store: ChatAgent) -> String {
        // Tauri parity: each result carries conversation context and the
        // chapter's global 1-based index (across all conversations), so the
        // model can feed it straight back to fetch_chapter_messages.
        let allChapters = store.allChapters
        let results: [[String: Any]] = chapters.map { ch in
            let conv = store.conversations.first { conv in
                conv.chapters.contains { $0.id == ch.id }
            }
            let globalIndex: Any = allChapters.firstIndex(where: { $0.id == ch.id })
                .map { $0 + 1 } ?? NSNull()
            return [
                "conversationId": conv?.id.uuidString ?? NSNull(),
                "conversationTitle": conv?.title ?? NSNull(),
                "chapter": globalIndex,
                "title": ch.title,
                "summary": String(ch.summary.prefix(240)),
                "keywords": ch.keywords,
            ]
        }

        guard let data = try? JSONSerialization.data(withJSONObject: results),
              let str = String(data: data, encoding: .utf8) else {
            return "[]"
        }
        return str
    }
}

// MARK: - Tool Definition Schema

struct ToolDef: Encodable {
    var type = "function"
    let function: FunctionDef

    struct FunctionDef: Encodable {
        let name: String
        let description: String
        let parameters: Parameters

        struct Parameters: Encodable {
            var type = "object"
            let properties: [String: Property]
            let required: [String]?
        }

        struct Property: Encodable {
            let type: String
            let description: String
        }
    }

    static let trackPerson = ToolDef(function: .init(
        name: "track_person",
        description: "查询一个人是否在历史对话中出现过，返回跨对话关系档案。当前对话中提到一个具体人名或身份（如'前任''老板''我妈'）时调用。",
        parameters: .init(properties: [
            "name": .init(type: "string", description: "要查询的人名或身份称呼"),
        ], required: ["name"])
    ))

    static let emotionTimeline = ToolDef(function: .init(
        name: "emotion_timeline",
        description: "返回用户最近N轮对话的情绪轨迹（原始数据，含情绪标签、强度和原文片段）。由模型自主判断趋势。当需要了解用户近期整体情绪状态时调用。",
        parameters: .init(properties: [
            "count": .init(type: "integer", description: "查询最近多少轮的情绪，默认5"),
        ], required: nil)
    ))

    static let searchChapters = ToolDef(function: .init(
        name: "search_chapters",
        description: "语义搜索历史章节（关键词 + 语义理解），返回标题、摘录和关键词。如需阅读完整原文请调 fetch_chapter_messages。当用户提到过去的某个话题、事件或关键词时调用。",
        parameters: .init(properties: [
            "query": .init(type: "string", description: "搜索关键词，留空返回最近5章"),
            "count": .init(type: "integer", description: "返回条数，默认5"),
        ], required: ["query"])
    ))

    static let fetchChapterMessages = ToolDef(function: .init(
        name: "fetch_chapter_messages",
        description: "获取指定章节的全部原文消息。当 search_chapters 返回的摘要不够详细、需要阅读完整原文时调用。",
        parameters: .init(properties: [
            "index": .init(type: "integer", description: "章节序号（从1开始），在章节索引中可看到章节编号"),
        ], required: ["index"])
    ))

    static let searchMemory = ToolDef(function: .init(
        name: "search_memory",
        description: "语义搜索跨对话记忆库（关键词 + 语义理解），返回相关的叙事摘要和洞察。当用户提到过去讨论过的话题、想回顾之前分析过的模式、或需要跨对话上下文时调用。这是长期记忆，不同于 search_chapters 只查当前对话。",
        parameters: .init(properties: [
            "query": .init(type: "string", description: "搜索关键词或短语，留空返回最近记忆"),
            "count": .init(type: "integer", description: "返回条数，默认10"),
        ], required: ["query"])
    ))

    static let manageNarrativeTimeline = ToolDef(function: .init(
        name: "manage_narrative_timeline",
        description: "读取或写入用户叙述中事件实际发生的时间轴，不使用消息发送时间。用户给出明确的具体日期、时间段（如高二期间）或两者混合时使用。若时间归属不清且会影响排序，先向用户提一个澄清问题，不要写入。补充旧事件前先 action=list，再用 eventId 更新。",
        parameters: .init(properties: [
            "action": .init(type: "string", description: "list 或 upsert"),
            "eventId": .init(type: "string", description: "更新已有节点时填写 list 返回的 ID；新增留空"),
            "title": .init(type: "string", description: "事件标题，最多24字"),
            "summary": .init(type: "string", description: "去重后的简短事实摘要，最多80字"),
            "startLabel": .init(type: "string", description: "用户表达的开始时间或单一时间节点，如 高二期间、2024年3月12日"),
            "endLabel": .init(type: "string", description: "可选结束时间；单点事件留空"),
            "timeKind": .init(type: "string", description: "period、date 或 mixed"),
            "sortIndex": .init(type: "integer", description: "按经历先后排序的整数，建议10、20、30；插入旧事件可用15等"),
        ], required: ["action"])
    ))
}

File: Prism/Tauri Version/src-tauri/src/tools.rs
use crate::archives::Archives;
use crate::models::{ChatRole, Conversation, NarrativeEvent};
use crate::search_expander::SearchExpander;
use serde_json::Value;
use uuid::Uuid;

pub struct ToolDef {
    pub name: &'static str,
    pub description: &'static str,
    pub parameters: Value,
}

pub fn tool_definitions() -> Vec<ToolDef> {
    vec![
        ToolDef {
            name: "track_person",
            description: "Look up person records across conversations",
            parameters: serde_json::json!({
                "type": "object",
                "properties": {
                    "name": {"type": "string", "description": "Person name to look up"}
                },
                "required": ["name"]
            }),
        },
        ToolDef {
            name: "emotion_timeline",
            description: "Return recent emotional trajectory",
            parameters: serde_json::json!({
                "type": "object",
                "properties": {
                    "count": {"type": "integer", "description": "Number of recent entries", "default": 5}
                },
                "required": []
            }),
        },
        ToolDef {
            name: "search_chapters",
            description: "Search historical chapter summaries by keywords",
            parameters: serde_json::json!({
                "type": "object",
                "properties": {
                    "query": {"type": "string", "description": "Search query; empty returns recent chapters"},
                    "count": {"type": "integer", "description": "Maximum results", "default": 10}
                },
                "required": ["query"]
            }),
        },
        ToolDef {
            name: "fetch_chapter_messages",
            description: "Get full messages for a specific chapter",
            parameters: serde_json::json!({
                "type": "object",
                "properties": {
                    "index": {"type": "integer", "description": "Global chapter index (1-based)"}
                },
                "required": ["index"]
            }),
        },
        ToolDef {
            name: "search_memory",
            description: "Long-term cross-conversation memory search",
            parameters: serde_json::json!({
                "type": "object",
                "properties": {
                    "query": {"type": "string", "description": "Search query"},
                    "count": {"type": "integer", "description": "Maximum results", "default": 10}
                },
                "required": ["query"]
            }),
        },
        narrative_timeline_tool(),
    ]
}

fn narrative_timeline_tool() -> ToolDef {
    ToolDef {
        name: "manage_narrative_timeline",
        description: "List or upsert events using the time stated in the user's story, never the chat send time. Use list before updating an earlier event. If the intended period is ambiguous, ask the user instead of calling upsert.",
        parameters: serde_json::json!({
            "type": "object",
            "properties": {
                "action": {"type": "string", "enum": ["list", "upsert"]},
                "eventId": {"type": "string", "description": "Existing event UUID when supplementing or correcting a past event"},
                "title": {"type": "string", "description": "Short event title"},
                "summary": {"type": "string", "description": "One concise non-repetitive sentence, at most 80 Chinese characters or 45 English words"},
                "startLabel": {"type": "string", "description": "User-stated start point or period, e.g. 高二期间 or 2024-03-12"},
                "endLabel": {"type": "string", "description": "Optional user-stated end point"},
                "timeKind": {"type": "string", "enum": ["period", "date", "mixed"]},
                "sortIndex": {"type": "integer", "description": "Chronological order; use gaps such as 10, 20, 30"}
            },
            "required": ["action"]
        }),
    }
}

pub fn narrative_timeline_tool_json() -> Vec<Value> {
    let tool = narrative_timeline_tool();
    vec![serde_json::json!({
        "type": "function",
        "function": {
            "name": tool.name,
            "description": tool.description,
            "parameters": tool.parameters
        }
    })]
}

pub fn tool_definitions_json() -> Vec<Value> {
    tool_definitions()
        .into_iter()
        .map(|t| {
            serde_json::json!({
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.parameters
                }
            })
        })
        .collect()
}

pub async fn execute_tool(
    name: &str,
    args: &Value,
    archives: &Archives,
    conversations: &[Conversation],
    current_conversation_id: Uuid,
) -> Result<Value, String> {
    match name {
        "track_person" => {
            let name = args["name"].as_str().unwrap_or("");
            let person = archives.find_person(name);
            Ok(serde_json::json!(person))
        }
        "emotion_timeline" => {
            let limit = args["count"].as_u64().unwrap_or(5).clamp(1, 50) as usize;
            let emotions = archives.get_recent_emotions(limit);
            Ok(serde_json::json!({"entries": emotions}))
        }
        "search_chapters" => {
            let query = args["query"].as_str().unwrap_or("");
            let limit = args["count"].as_u64().unwrap_or(10).clamp(1, 20) as usize;
            let q = query.to_lowercase();
            let terms = SearchExpander::default().terms(&q);
            let mut scored = Vec::new();

            for (conversation_index, conversation) in conversations.iter().enumerate() {
                for (chapter_index, chapter) in conversation.chapters.iter().enumerate() {
                    let title = chapter.title.to_lowercase();
                    let summary = chapter.summary.to_lowercase();
                    let keywords = chapter
                        .keywords
                        .iter()
                        .map(|keyword| keyword.to_lowercase())
                        .collect::<Vec<_>>();
                    let mut score = if q.trim().is_empty() { 1 } else { 0 };
                    for term in terms.iter().filter(|term| !term.is_empty()) {
                        if title.contains(term) {
                            score += 3;
                        }
                        if keywords.iter().any(|keyword| keyword.contains(term)) {
                            score += 2;
                        }
                        if summary.contains(term) {
                            score += 1;
                        }
                    }
                    if !q.trim().is_empty() && title.contains(&q) {
                        score += 10;
                    }
                    if !q.trim().is_empty() && summary.contains(&q) {
                        score += 3;
                    }
                    if score > 0 {
                        scored.push((
                            score,
                            conversation_index,
                            chapter_index,
                            serde_json::json!({
                                "conversationId": conversation.id,
                                "conversationTitle": conversation.title,
                                "chapter": chapter_index + 1,
                                "title": chapter.title,
                                "summary": chapter.summary.chars().take(240).collect::<String>(),
                                "keywords": chapter.keywords,
                            }),
                        ));
                    }
                }
            }

            scored.sort_by(|left, right| {
                right
                    .0
                    .cmp(&left.0)
                    .then_with(|| left.1.cmp(&right.1).then_with(|| left.2.cmp(&right.2)))
            });
            Ok(serde_json::json!({
                "query": q,
                "results": scored
                    .into_iter()
                    .take(limit)
                    .map(|entry| entry.3)
                    .collect::<Vec<_>>()
            }))
        }
        "fetch_chapter_messages" => {
            let index = args["index"].as_i64().unwrap_or(0);
            if index < 1 {
                return Ok(serde_json::json!({"error": "invalid index"}));
            }

            let mut current = 0_i64;
            for conversation in conversations {
                for chapter in &conversation.chapters {
                    current += 1;
                    if current != index {
                        continue;
                    }
                    let messages = conversation
                        .messages
                        .iter()
                        .filter(|message| {
                            message.role != ChatRole::system
                                && chapter.message_ids.contains(&message.id)
                        })
                        .take(12)
                        .map(|message| {
                            serde_json::json!({
                                "role": message.role,
                                "content": message.content,
                            })
                        })
                        .collect::<Vec<_>>();
                    return Ok(serde_json::json!({
                        "chapter": index,
                        "title": chapter.title,
                        "summary": chapter.summary,
                        "keywords": chapter.keywords,
                        "messages": messages,
                    }));
                }
            }
            Ok(serde_json::json!({
                "error": "chapter not found",
                "max": current,
                "currentConversationId": current_conversation_id,
            }))
        }
        "search_memory" => {
            let query = args["query"].as_str().unwrap_or("");
            let limit = args["count"].as_u64().unwrap_or(10).clamp(1, 20) as usize;
            let results = archives
                .search_memory(&[query.to_string()])
                .into_iter()
                .take(limit)
                .map(|entry| {
                    serde_json::json!({
                        "content": entry.content,
                        "keywords": entry.keywords,
                        "sourceChapter": entry.source_chapter_title,
                        "timeSpanStart": entry.time_span_start.map(|value| value.to_rfc3339()),
                        "timeSpanEnd": entry.time_span_end.map(|value| value.to_rfc3339()),
                        "recallCount": entry.recall_count,
                    })
                })
                .collect::<Vec<_>>();
            Ok(serde_json::json!(results))
        }
        "manage_narrative_timeline" => {
            let action = args["action"].as_str().unwrap_or("list");
            let current_events = || {
                archives
                    .all_narrative_events()
                    .into_iter()
                    .filter(|event| event.conversation_id == current_conversation_id)
                    .collect::<Vec<_>>()
            };
            if action == "list" {
                return Ok(serde_json::json!({"events": current_events()}));
            }
            if action != "upsert" {
                return Err("action must be list or upsert".to_string());
            }

            let title = args["title"].as_str().unwrap_or("").trim();
            let start_label = args["startLabel"].as_str().unwrap_or("").trim();
            if title.is_empty() || start_label.is_empty() {
                return Err("title and startLabel are required; ask the user when the story time is unclear".to_string());
            }
            let requested_id = args["eventId"]
                .as_str()
                .and_then(|value| Uuid::parse_str(value).ok());
            let existing = requested_id
                .and_then(|id| current_events().into_iter().find(|event| event.id == id));
            let source_message_id = conversations
                .iter()
                .find(|conversation| conversation.id == current_conversation_id)
                .and_then(|conversation| {
                    conversation
                        .messages
                        .iter()
                        .rev()
                        .find(|message| message.role == ChatRole::user)
                        .map(|message| message.id)
                });
            let mut source_message_ids = existing
                .as_ref()
                .map(|event| event.source_message_ids.clone())
                .unwrap_or_default();
            if let Some(message_id) = source_message_id {
                if !source_message_ids.contains(&message_id) {
                    source_message_ids.push(message_id);
                }
            }
            let summary = args["summary"]
                .as_str()
                .unwrap_or("")
                .trim()
                .chars()
                .take(160)
                .collect::<String>();
            let event = NarrativeEvent {
                id: existing
                    .as_ref()
                    .map(|event| event.id)
                    .unwrap_or_else(Uuid::new_v4),
                conversation_id: current_conversation_id,
                title: title.chars().take(40).collect(),
                summary,
                start_label: start_label.chars().take(60).collect(),
                end_label: args["endLabel"]
                    .as_str()
                    .unwrap_or("")
                    .trim()
                    .chars()
                    .take(60)
                    .collect(),
                time_kind: args["timeKind"]
                    .as_str()
                    .filter(|value| matches!(*value, "period" | "date" | "mixed"))
                    .unwrap_or("period")
                    .to_string(),
                sort_index: args["sortIndex"]
                    .as_i64()
                    .map(|value| value.clamp(-100_000, 100_000) as i32)
                    .or_else(|| existing.as_ref().map(|event| event.sort_index))
                    .unwrap_or_else(|| {
                        current_events()
                            .iter()
                            .map(|event| event.sort_index)
                            .max()
                            .unwrap_or(0)
                            + 10
                    }),
                source_message_ids,
                updated_at: chrono::Utc::now(),
            };
            archives.upsert_narrative_event(event.clone());
            Ok(serde_json::json!({"saved": true, "event": event}))
        }
        _ => Err(format!("Unknown tool: {}", name)),
    }
}
